packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js

```diff
@@ -0,0 +1,463 @@
+  const editor = useSlate();
+  const { components } = useLibrary();
+  const MediaLibraryDialog = components['media-library'];
+
+  const insertImages = (images) => {
+    images.forEach((img) => {
+      const image = { type: 'image', image: img, children: [{ type: 'text', text: '' }] };
+      Transforms.insertNodes(editor, image);
+    });
+  };
+
+  const handleSelectAssets = (images) => {
+    const formattedImages = images.map((image) => {
+      // create an object with imageSchema defined and exclude unnecessary props coming from media library config
+      const expectedImage = pick(image, IMAGE_SCHEMA_FIELDS);
+
+      return {
+        ...expectedImage,
+        alternativeText: expectedImage.alternativeText || expectedImage.name,
+        url: prefixFileUrlWithBackendUrl(image.url),
+      };
+    });
+
+    insertImages(formattedImages);
+    handleClose();
+  };
+
+  return (
+    <MediaLibraryDialog
+      allowedTypes={[ALLOWED_MEDIA_TYPE]}
+      onClose={handleClose}
+      onSelectAssets={handleSelectAssets}
+    />
+  );
+};
+
+ImageDialog.propTypes = {
+  handleClose: PropTypes.func.isRequired,
+};
+
+const isLastBlockImageOrCode = (editor) => {
+  const { selection } = editor;
+
+  if (!selection) return false;
+
+  const [currentImageORCodeBlock] = Editor.nodes(editor, {
+    at: selection,
+    match: (n) => n.type === 'image' || n.type === 'code',
+  });
+
+  if (currentImageORCodeBlock) {
+    const [, currentNodePath] = currentImageORCodeBlock;
+
+    const isNodeAfter = Boolean(Editor.after(editor, currentNodePath));
+
+    return !isNodeAfter;
+  }
+
+  return false;
+};
+
+export const BlocksDropdown = () => {
+  const editor = useSlate();
+  const { formatMessage } = useIntl();
+  const [isMediaLibraryVisible, setIsMediaLibraryVisible] = React.useState(false);
+
+  const blocks = useBlocksStore();
+  const blockKeysToInclude = Object.entries(blocks).reduce((currentKeys, entry) => {
+    const [key, block] = entry;
+
+    return block.isInBlocksSelector ? [...currentKeys, key] : currentKeys;
+  }, []);
+
+  const [blockSelected, setBlockSelected] = React.useState(Object.keys(blocks)[0]);
+
+  /**
+   * @param {string} optionKey - key of the heading selected
+   */
+  const selectOption = (optionKey) => {
+    toggleBlock(editor, blocks[optionKey].value);
+
+    setBlockSelected(optionKey);
+
+    if (isLastBlockImageOrCode(editor)) {
+      // insert blank line to add new blocks below code or image blocks
+      Transforms.insertNodes(
+        editor,
+        {
+          type: 'paragraph',
+          children: [{ type: 'text', text: '' }],
+        },
+        { at: [editor.children.length] }
+      );
+    }
+
+    if (optionKey === 'image') {
+      setIsMediaLibraryVisible(true);
+    }
+  };
+
+  return (
+    <>
+      <Select
+        startIcon={<Icon as={blocks[blockSelected].icon} />}
+        onChange={selectOption}
+        placeholder={blocks[blockSelected].label}
+        value={blockSelected}
```

Comment:
How come you need to control the state of this dropdown? Is it manipulated externally? 🤔

Reply:
Yes. I need to change the option selected everytime I select a portion of content inside the RTE editor, for example
if I select a portion of text that is an h3 I need to update immediately the option selected in the Dropdown to Heading 3
<img width="716" alt="selection" src="https://github.com/strapi/strapi/assets/2589748/727f34ee-6bfe-47f5-8ef4-06ce68f9d658">